src/lib/CofEnv.ml

```diff
@@ -67,58 +120,57 @@ struct
-      left local (psis @ cx) phi
-end
+  module Seq = struct type t = bool let vacuous = true let seq = List.for_all end
+  module M = Search (Seq)
 
-let test env phi =
-  match env.status with
-  | `Inconsistent ->
-    true
-  | `Consistent ->
-    let local = Inversion.{classes = env.classes; true_vars = env.true_vars} in
-    Inversion.left local [env.cof] phi
+  let sequent env cx phi = M.left_invert' env cx (fun env -> right env phi)
+  let simple env phi = sequent env [] phi
+end
 
-let test_sequent (env : env) cx phi =
-  let local = Inversion.{classes = env.classes; true_vars = env.true_vars} in
-  Inversion.left local (env.cof :: cx) phi
+let test_sequent env cx phi =
+  match env with
+  | `Inconsistent -> true
+  | `Consistent env -> Test.sequent env cx phi
 
-let rec assume env phi =
-  match env.status with
+let assume env phi =
+  match env with
   | `Inconsistent -> env
-  | `Consistent ->
-    let phi = Cof.reduce phi in
-    match phi with
-    | Cof.Var v ->
-      {env with true_vars = VarSet.add v env.true_vars}
-    | Cof.Cof phi ->
-      match phi with
-      | Cof.Meet phis ->
-        List.fold_left assume env phis
-      | Cof.Join _ ->
-        if test_sequent env [Cof.Cof phi] Cof.bot then inconsistent else
-          {env with cof = Cof.meet [env.cof; Cof.Cof phi]}
-      | Cof.Eq (r, s) ->
-        let classes = UF.union r s env.classes in
-        if UF.find D.Dim0 classes = UF.find D.Dim1 classes then
-          inconsistent
-        else
-          {env with classes}
-
-let equate env r s =
-  assume env @@ Cof.eq r s
+  | `Consistent env ->
+    let rec go env =
+      function
+      | [] -> `Consistent env
+      | (phi :: phis) ->
+        match phi with
+        | Cof.Var v ->
+          go {env with true_vars = VarSet.add v env.true_vars} phis
+        | Cof.Cof phi ->
+          match phi with
+          | Cof.Meet psis ->
+            go env (psis @ phis)
+          | Cof.Join psis ->
+            let env = {env with unreduced_joins = psis :: env.unreduced_joins} in
+            if Test.simple env Cof.bot then `Inconsistent else go env phis
+          | Cof.Eq (r, s) ->
+            let classes = UF.union r s env.classes in
+            if UF.find D.Dim0 classes = UF.find D.Dim1 classes then
+              `Inconsistent
+            else
+              go {env with classes} phis
+    in go env [Cof.reduce phi] (* do we want Cof.reduce here? *)
+
+(** Monadic interface *)
+module M (M : CoolBasis.Monad.S) :
+sig
+  (* Search all branches under one more cofibration *)
+  val left_invert_under_cofs : env -> D.cof list -> (env -> unit M.m) -> unit M.m
+end
+=
+struct
+  module MU = CoolBasis.Monad.Util (M)
+  module Seq = struct type t = unit M.m let vacuous = M.ret () let seq = MU.app end
+  module S = Search (Seq)
+
+  let left_invert_under_cofs env phis cont =
+    S.left_invert env phis @@ fun {classes; true_vars} ->
+      cont @@ `Consistent {classes; true_vars; unreduced_joins = []}
```

Comment:
Does it look better now?

Reply:
well, i wouldn’t indent the next line. in case it helps, i generally follow the style produced by ocp-indent